src/environmentd/src/http.rs

```diff
@@ -182,8 +190,359 @@ struct AuthedUser {
+pub struct SqlHttpExecuteResponse {
+    pub results: Vec<SimpleResult>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ExtendedRequest {
+    // TODO: we can remove this alias once platforms issues these requests
+    // using `query`.
+    #[serde(alias = "sql")]
+    /// 0 or 1 queries
+    pub query: String,
+    /// Optional parameters for the query
+    #[serde(default)]
+    pub params: Vec<Option<String>>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum HttpSqlRequest {
+    /// 0 or more queries delimited by semicolons in the same strings. Do not
+    /// support parameters.
+    Simple {
+        #[serde(alias = "sql")]
+        query: String,
+    },
+    /// 0 or more [`ExtendedRequest`]s.
+    Extended { queries: Vec<ExtendedRequest> },
+}
+
 pub struct AuthedClient(pub SessionClient);
 
+impl AuthedClient {
+    /// Executes an [`HttpSqlRequest`].
+    ///
+    /// The provided `stmts` are executed directly, and their results
+    /// are returned as a vector of rows, where each row is a vector of JSON
+    /// objects.
+    pub async fn execute_sql_http_request(
+        &mut self,
+        request: HttpSqlRequest,
+    ) -> Result<SqlHttpExecuteResponse, anyhow::Error> {
+        let mut results = vec![];
+
+        match request {
+            HttpSqlRequest::Simple { query } => {
+                let stmts = mz_sql::parse::parse(&query).map_err(|e| anyhow!(e))?;
+                let num_stmts = stmts.len();
+
+                for stmt in stmts {
+                    if matches!(self.0.session().transaction(), TransactionStatus::Failed(_)) {
+                        break;
+                    }
+                    // Mirror the behavior of the PostgreSQL simple query protocol.
+                    // See the pgwire::protocol::StateMachine::query method for details.
+                    if let Err(e) = self.0.start_transaction(Some(num_stmts)).await {
+                        results.push(SimpleResult::err(e));
+                        break;
+                    }
+                    let res = self.execute_stmt(stmt, vec![]).await;
+                    if matches!(res, SimpleResult::Err { .. }) {
+                        self.0.fail_transaction();
+                    }
+                    results.push(res);
+                }
+            }
+            HttpSqlRequest::Extended { queries } => {
+                for ExtendedRequest { query, params } in queries {
+                    if matches!(self.0.session().transaction(), TransactionStatus::Failed(_)) {
+                        break;
+                    }
+
+                    let mut stmts = match mz_sql::parse::parse(&query) {
+                        Ok(stmts) => stmts,
+                        Err(e) => {
+                            results.push(SimpleResult::err(e));
+                            break;
+                        }
+                    };
+
+                    if stmts.len() > 1 {
```

Comment:
We should change this to be exactly one, and error on zero.

Reply:
Those semantics differ from the Extended Query protocol:

> Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages: CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended.

This suggests that the `EmptyQueryResponse` is distinct from the `ErrorResponse`.

I'm also not able to replicate the behavior you suggest in `pgtest`, so put up #14774 to expand the test suite to cover this case. Looks like the `EmptyQueryResponse` is not treated as an error in either the simple or extended protocols, i.e. subsequent commands continue to execute.

To express this to the client of this protocol, what do you suggest? Should we add a new response type for the empty query rather than simply eliding it?

Reply:
Ah I forgot that extended supports executing the empty message. I'm ok with either supporting it or not, although I have a light preference to not supporting the empty message, because then we don't have to figure out how to send a "empty statement" response back.

Yes, an empty statement is not an error in pg (definitely in simple, and I forgot also in extended) and shouldn't stop execution.

Reply:
> To express this to the client of this protocol, what do you suggest? Should we add a new response type for the empty query rather than simply eliding it?

I suggest we error on empty statement here, even though it's a difference from the extended protocol. In practice the empty statement is never sent over extended, and I'm not sure why we would need to support it over http. If we need to, we can support it and add a response type for it.

Reply:
Sure thing. Will be updated on next push.